experiment/popularity_community/crawl_torrents.py

```diff
@@ -0,0 +1,217 @@
+
+### Example
+
+```
+python3 crawl_torrents.py -t 600
+python3 crawl_torrents.py -t 600 -f result.json
+```
+
+"""
+import argparse
+import asyncio
+import json
+import logging
+import os
+from collections import defaultdict
+
+import lz4
+
+from ipv8.lazy_community import lazy_wrapper
+from ipv8.peer import Peer
+from ipv8.peerdiscovery.discovery import RandomWalk
+from tool.tiny_tribler_service import TinyTriblerService
+from tribler_core.modules.metadata_store.community.remote_query_community import RemoteQueryCommunity, \
+    SelectResponsePayload, SelectRequest, RemoteSelectPayload
+from tribler_core.modules.metadata_store.serialization import REGULAR_TORRENT
+from tribler_core.utilities.unicode import hexlify
+
+UNLIMITED = -1  # this magic number is use inside many Tribler modules
+
+IPV8_WALK_INTERVAL = 0.05
+
+RANDOM_WALK_TIMEOUT_IN_SEC = 5
+RANDOM_WALK_WINDOW_SIZE = 0
+RANDOM_WALK_RESET_CHANCE = 30
+
+CRAWLER_REQUEST_TIMEOUT_IN_SEC = 60
+CRAWLER_SAVE_INTERVAL_IN_SEC = 60
+
+
+class CrawlerSelectRequest(SelectRequest):
+    """Custom class for RemoteQueryCommunity select request
+
+    It has been added only for having an ability to increase `timeout_delay`
+    """
+
+    @property
+    def timeout_delay(self):
+        return CRAWLER_REQUEST_TIMEOUT_IN_SEC
+
+
+class Crawler(RemoteQueryCommunity):
+    """ Crawler based on RemoteQueryCommunity
+
+    The approach based on two ideas:
+        * use "aggressive" values for network walking (see ExperimentConfiguration.RANDOM_WALK_*)
+        * use RemoteQueryCommunity's ability to select MetadataInfo from remote db
+    """
+
+    def __init__(self, my_peer, endpoint, network, metadata_store, out_file_path):
+        super().__init__(my_peer, endpoint, network, metadata_store)
+        self._logger = logging.getLogger(self.__class__.__name__)
+        self._output_file_path = out_file_path
+        self.max_peers = UNLIMITED
+        self.queried_peers_limit = UNLIMITED
+
+        self.torrent_dictionary = defaultdict(list)  # {"Peer": (votes, infohash, title)}
+        self.polled_peers = set()
+
+        self.register_task("request_torrents", self.request_torrents_from_new_peers, interval=10)
+        self.register_task("save_torrents", self.save, interval=CRAWLER_SAVE_INTERVAL_IN_SEC)
+
+    def __del__(self):
+        self.save()
+
+    def request_torrents_from_new_peers(self):
+        """Send a request for gathering torrent information
+
+        Behaviour:
+
+        For all new pears: ```self.get_peers() - self.polled_peers```
+        the crawler request 100 most popular torrents.
+
+        Note:   it is possible to request more, but it requires a significant
+                increase of code complexity.
+        Ref: https://github.com/Tribler/tribler/blob/devel/src/tribler-core/tribler_core/modules/metadata_store/community/remote_query_community.py#L19
+        """
+        peers_ready_to_request = set(self.get_peers()) - self.polled_peers
```

Comment:
@qstokkink , is it the best way to query peers aggressively?

@grimadas , you've built an aggressive crawler some time ago. Did you use something like this?

Reply:
@ichorid Well, it is definitely very aggressive for the peers you know every 10 seconds. However, if the intent of this thing is to actually _crawl_, you'll also want to call `self.network.remove_peer()` to get rid of peers you the have collected information for (the ones in `self.polled_peers`) to get some new peers.

Reply:
One disclaimer: setting the `max_peers` to infinity will also eventually yield you all of the information, just slower. So, is this the most aggressive? No. Will it work? Yes.

Reply:
@qstokkink cool! thank you for `self.network.remove_peer()` tip.